### ISSUE TITLE
[Package Request] Snap package

### ISSUE THREAD

**Issue by extraymond** [zotero/zotero#1288]
Hi! Thx for creating zotero.

I would like to know if it's possible to add snap package, so we can deliver zotero and updating it automatically.

Flatpak is good option too.

**Comment by ibaidev** [comment]
I am trying to create a snap package with a code I've found in the snapcraft forum:
https://forum.snapcraft.io/t/zotero-snap-failing-with-version-glibc-private-not-defined-in-file-libc-so-6/2463/8

I started the following project:
https://github.com/ibaidev/zotero-snap

I use https://build.snapcraft.io/ to build the snap and https://dashboard.snapcraft.io/ to publish in the store. That script downloads the last tar.gz for linux-x86_64 available in https://www.zotero.org/download/client/dl?channel=release&platform=linux-x86_64 to build the snap. I have not worked in other architectures yet. If a commit is pushed to my github repository it triggers the build. It can be also manually triggered, and always downloads from that URL.

It can be installed with:
sudo snap install --edge zotero-standalone

It runs well with "/snap/zotero-standalone/current/zotero" but I guess that it runs out of the snap sandbox. When I try to run "snap run zotero-standalone" I get a "Bad system call". I am working to fix this. Any help is welcome.

My proposal is to have the "snap/" folder in the zotero official repository (or a separate repository owned by zotero) so that when a new zotero version is released a new snap is built.

**Comment by dstillman** [comment]
Same question I asked in your [Zotero Forums thread](https://forums.zotero.org/discussion/69158/snap-package-for-zotero): what happens if there's an update on the regular release channel? Would we need/want to disable the built-in update checking? If we rely on the snap update mechanism, how well do the delta updates work? For comparison, our delta update for 5.0.28->5.0.29 was 147KB, vs. 53MB for the full tarball.

Also, it seems like the <a href="https://github.com/ibaidev/zotero-snap/blob/a90ccb43eb2db267d48cb1193b3fb19b40545eb9/snap/snapcraft.yaml#L14">defined permissions</a> (if I'm understanding this right) could be pared down quite a bit further, no?

**Comment by ibaidev** [comment]
https://build.snapcraft.io/ creates a webhook in the github repository. This webhook can be disabled and the builds can be manually triggered. It depends on how you want to manage the deployment.

I haven't tested the delta updates in depth. As the changes between my versions are small, very few files are downloaded each update.

Yes, probably I'm working with too many permissions. If I make "snap run zotero-standalone" work, I will try to remove the unnecessary permissions.

**Comment by ibaidev** [comment]
That "Bad system call" seems to be conflict between the snapd's sandbox and the firefox's sandbox. We should either disable the firefox's sandbox or ask snapd to allow it. I was not able to disable it, as people from chromium/brave/electron does, so I asked the permission to snapd. Although I was able to create the snap, install it on my computer and it works X11, it fails with wayland. There is a "Segmentation fault" that I am not able to solve. Moreover, as I ask the permission to run a snap with a sandbox on it, It must be manually reviewed in order to be published in snapcraft.

Is there someone that knows how to disable the firefox sandbox? Otherwise, are there any plans to migrate zotero to electron? It seems that it would be easier to snap.

**Comment by eebrown** [comment]
In case you weren't aware, there is an up-to-date Zotero flatpak on flathub.
https://flathub.org/apps/details/org.zotero.Zotero
This may be sufficient and if so this issue could be closed.

**Comment by extraymond** [comment]
Well! Things have gone full cycle.
Turns out I can make this happen after all these years.

Here's the link if anyone is interested. Would love to get some feedback on the snap.
[zotero-snap](https://snapcraft.io/zotero-snap)

**Comment by extraymond** [comment]
Are there rectangular icons for Zotero? Snapcraft required one to be used.

**Comment by extraymond** [comment]
Hi! I've created a confined [zotero-snap](https://snapcraft.io/zotero-snap), if the organization wants it I would love to transfer the snap's ownership to you.

**Comment by dstillman** [comment]
@extraymond: I'd be happy to look at this, though I don't know much about snaps, so I'd want to find out more about how this is created and any differences from a stock install before we agree to take it over. Are there any restrictions to the snap that would make it not behave like normal Zotero? Does a confined snap not prevent normal file access (adding files from the filesystem, creating linked files with ZotFile, etc.) from working?

**Comment by extraymond** [comment]
@dstillman Thx!

**general snap stuffs**

Sure! Snaps are created in a build -> stage -> prime -> snap pipeline. A snap can be build from various parts too. 
In the build step the build-packages will be installed from apt to fulfill the build requirement and will not get copied into the next stage. In the stage step, the stage-packages will pull the runtime required packages from repo and copy them into the snapped filesystem. Lastly the prime step, if you wan to change the installation directory or dump in some files, you can achieve that in the prime step.

So the contract in those metadata will make sure the application is build with the exact needed libraries. This is typically how snap's are build. It will be self contained and not allowed to access the outside world if allowed.

As of how applications are calling path or access system resources, it's using interfaces(slots/plugs) to interact. There are some defined interfaces for a typical desktop-app the we can insert into the plugs field to allow the app to ask for permission.

---

**zotero related**

Speaking about the snap I created, there's a dump plugin to just copy the release packages without building it from scratch. So I only adjust the launcher icon and make the required libraries included. 

As of the plugs, network-bind and home obvious, browser-support is to allow access to /dev/shm to let the app maintain sandbox on its own. Otherwise snaps can't access /dev/shm. And lastly removable-media to access mount directory.

Although I haven't read the source code of zotero, if it's using freedesktop standards things would be quite seamless. Just a few places it would need attention.

Snaps have two important path in runtime, $SNAP_USER_DATA and $SNAP_USER_COMMON.
The first one will get restored or backup if reverting. The other one will stay the same regardless of version upgrade. $SNAP_USER_COMMON. $HOME and $XDG_CONFIG_HOME etc will point to $SNAP_USER_COMMON/*.

As of linking, if your linking get crossed the allowed boundaries, it will get rejected. For example, ln -s /mount/path target without allowing removable-media will get ejected.

As of system file access inside the snap.If there are hardcoded system paths, we can use layout to redirect /usr/lib to $SNAP/usr/lib so it can find the correct path without touching the app codebase.

---

I believe there are some upgrade services zotero does when upgrading. If this app is upgraded by the snap daemon, I would suggest that the upgrade operation have good access to the correct paths like mentioned above. I encounter an error when bumping snap version with a existing $SNAP_USER_COMMON saying reading sqlite file failed.

**Comment by extraymond** [comment]
Relevant links:
For accessing system/user resources.
[snaps interfaces docs](https://snapcraft.io/docs/supported-interfaces)

For accessing other app's(snap or not) resources and allowing(other snaps) access to the snaps resource from outside.
[interface management](https://snapcraft.io/docs/interface-management)

[Useful ENV_VARIABLE](https://snapcraft.io/docs/environment-variables)

**Comment by dstillman** [comment]
To clarify, "linked files" in this context doesn't mean symlinks — it means Zotero [linked files](https://www.zotero.org/support/attaching_files#stored_files_and_linked_files), which can be anywhere on disk. Those can be added manually or they can be created automatically with the ZotFile plugin, which moves files to a configured location. And, actually, it's not limited to linked files, because even for stored files you can just drag a PDF on disk into Zotero to add it or use a filepicker.

Are you saying that Zotero would still have access to the user's entire home directory and removable disks, such that the above things would still work as long as the user didn't try to access system files? If so, what happens if the user accidentally navigates outside the home directory via the filepicker? Or is there a virtual filesystem such that other things don't show up?

As long as there's unrestricted access to those locations, that might be OK. But from [the interfaces page](https://snapcraft.io/docs/supported-interfaces) it seems like those require a forum post with a permission request, which I don't really want to get into. Or have you already done that for the existing package?

Some Zotero plugins might do other things, like run external programs. Would that fail?

Basically, I'm hesitant to get into supporting something that wouldn't be part of normal development and testing and that I imagine could fail in various ways to do changes in the sandbox, Zotero's behavior, or the frameworks Zotero is built on. There's an [open pull request](https://github.com/zotero/zotero-standalone-build/pull/73) for taking over @retorquere's deb packages, and because that's purely a wrapping of the tarball, it won't require any additional testing, so I'm much more comfortable with that approach.

**Comment by extraymond** [comment]
> To clarify, "linked files" in this context doesn't mean symlinks — it means Zotero [linked files](https://www.zotero.org/support/attaching_files#stored_files_and_linked_files), which can be anywhere on disk. Those can be added manually or they can be created automatically with the ZotFile plugin, which moves files to a configured location. And, actually, it's not limited to linked files, because even for stored files you can just drag a PDF on disk into Zotero to add it or use a filepicker.
> 
> Are you saying that Zotero would still have access to the user's entire home directory and removable disks, such that the above things would still work as long as the user didn't try to access system files? If so, what happens if the user accidentally navigates outside the home directory via the filepicker? Or is there a virtual filesystem such that other things don't show up?
> 
> As long as there's unrestricted access to those locations, that might be OK. But from [the interfaces page](https://snapcraft.io/docs/supported-interfaces) it seems like those require a forum post with a permission request, which I don't really want to get into. Or have you already done that for the existing package?
> 
> Some Zotero plugins might do other things, like run external programs. Would that fail?
> 
> Basically, I'm hesitant to get into supporting something that wouldn't be part of normal development and testing and that I imagine could fail in various ways due to changes in the sandbox, Zotero's behavior, or the frameworks Zotero is built on. There's an [open pull request](https://github.com/zotero/zotero-standalone-build/pull/73) for taking over @retorquere's deb packages, and because that's purely a wrapping of the tarball, it won't require any additional testing, so I'm much more comfortable with that approach.

**about file direcotry**

I've only tested about home directories an mount points but here's what you get when trying to access 
path without permission.

File picker will popup no permission dialogue.

![2019-12-29 17-37-15 的螢幕擷圖](https://user-images.githubusercontent.com/707176/71555048-1ecaf300-2a62-11ea-8ce5-87af4f9f5b46.png)

I add the zotero folder to my cloud sync services and it can be seen by bot the sync program and zotero.

![2019-12-29 17-37-52 的螢幕擷圖](https://user-images.githubusercontent.com/707176/71555049-238fa700-2a62-11ea-8e23-a6f9e091ba6f.png)

**about plugins**

I don't have much experience about plugins. But if plugins are installed in the permitted path and they have a fixed dependency on the program, then it would work. Do you have some suggestions about trying few?

**Comment by extraymond** [comment]
Dumping deb packages as the source of snap is common too. If that's aligned with the project's building pipeline, I suppose it wouldn't be too hard to track.

**Comment by retorquere** [comment]
> Some Zotero plugins might do other things, like run external programs.

Like BBT. BBT runs `git` and `texstudio` if so configured.

**Comment by dstillman** [comment]
Yeah, I don't think this is something we want to get into. Plugins can do whatever they want, and while a sandbox mechanism with officially approved overrides might make sense if it applied to all platforms, something that only affected a small number of users on Linux would just result in broken plugins — and quite possibly other things, unexpectedly, in the future — for those users.

So while we're still planning to officially support deb builds, I don't think snaps are for us.

Thanks for your effort on this, though, @extraymond.

**Comment by extraymond** [comment]
Well. I understand plugin system are important. There is actually another way for snaps.
Snaps can have classic confinement. Which works without the benefit of confinement.

Atom, Vscode and various app that needs system file access are confined this way.
I can try this route if it's a acceptable solution?

**Comment by dstillman** [comment]
But classic confinement seems to require us to request manual approval and for the user to pass the `--classic` command line flag to install the snap [1]. We're not interested in going through the former and the latter would seem to defeat much of the advantage of the snap over the deb repo.

[1] https://snapcraft.io/docs/snap-confinement

**Comment by extraymond** [comment]
About manual review, if it's directly come from upstream(you guys), but not me, snapcraft teams have fantastic response since they are pushing for the adoption.

About classic flag, if the user is using software store on their specific distro, such as gnome-software and kde discover, there are snap plugins so user won't need to go to command line to install the app.

Even without confinement, snaps have some advantages, as a user I liked those features the most:

1. Automatic delta updates, I will always get the latest version from developer.
2. Safe revert system, if the new release have bugs, users can snap revert and get the last working version. No need to remove and install older version. App data are backuped automatically too.
3. Channel selection, some times I like to try features in the beta version, so I can help discover bugs or report user experience to help devs make better software.

**Comment by extraymond** [comment]
And atom is just a deb dump classic snap, so you can still focus on maintaining the deb while having a way to deliver updates automatically to your user.

**Comment by dstillman** [comment]
> About manual review, if it's directly come from upstream(you guys), but not me, snapcraft teams have fantastic response since they are pushing for the adoption.

That may be, but it's still not a process we're particularly interested in going through. Can you link to examples where, say, Atom or Firefox requested permission and were approved?

> And atom is just a deb dump classic snap

What about Firefox? That'd be another close example on a technical level.

I can't guarantee we'll move forward with this, but we can look at whatever build materials you can provide as examples.

**Comment by goliatl** [comment]
I just installed the zotero snap and zotero cannot find the libreoffice extension, so this seems to be a problem

**Comment by extraymond** [comment]
> I just installed the zotero snap and zotero cannot find the libreoffice extension, so this seems to be a problem

Thx for reporting, I'll make another attempt to build the snap with classic confinement. That would make the snap works unconfined but allowing more flexibility for the extension framework zotero used.

**Comment by extraymond** [comment]
> > About manual review, if it's directly come from upstream(you guys), but not me, snapcraft teams have fantastic response since they are pushing for the adoption.
> 
> That may be, but it's still not a process we're particularly interested in going through. Can you link to examples where, say, Atom or Firefox requested permission and were approved?
> 
> > And atom is just a deb dump classic snap
> 
> What about Firefox? That'd be another close example on a technical level.
> 
> I can't guarantee we'll move forward with this, but we can look at whatever build materials you can provide as examples.

Firefox was build and maintained officially by mozilla. If the snap within the store it's provided by upstream, the publisher account would be granted verified badge. I've reporting the snap to grant permission to gamepad access plug on bugzilla, and it works.

I'll try to make a classic confined snap so it won't restrict the plugin. I think it works better for zotero at the moment.

I'll try and find more materials and see what I can do. Thx for your suggestion.

**Comment by hadim** [comment]
The snap package lags a few version behind: https://github.com/extraymond/zotero-snap/pull/36

Any chance a snap package could be part of the official zotero distribution?

**Comment by retorquere** [comment]
I'll see if I can incorporate it in my deb builder.

**Comment by extraymond** [comment]
> The snap package lags a few version behind: [extraymond/zotero-snap#36](https://github.com/extraymond/zotero-snap/pull/36)
> 
> Any chance a snap package could be part of the official zotero distribution?

> I'll see if I can incorporate it in my deb builder.

If this is referring to the official zotero deb, let me know what I need to do to transfer the package to the zotero owners.

Some info about the current snap build recipe:
1. the current package is build in a plain way that it download and unpack a deb and set the dependencies that the runtime requires.
2. it should be able to reproduce most of the build steps of deb builder since snap support various build tools and can just use plain shell command to drive the building process.

Let me know if there's anything I can help to make this happen.
I never found a convenient way to get in sync with the official release schedule since I don't use zotero as often as I used to be.

**Comment by retorquere** [comment]
My deb package contains the official binaries, but the packages themselves are not (yet) built/offered by Zotero.

That said, my build tests the packages by installing them, so if the snaps build on that, it would be simple to incorporate the snap builds. My packages update fully automatically every two hours, so deb releases are never more than two hours behind the official release. It's entirely hands-off.

**Comment by extraymond** [comment]
> My deb package contains the official binaries, but the packages themselves are not (yet) built/offered by Zotero.
> 
> That said, my build tests the packages by installing them, so if the snaps build on that, it would be simple to incorporate the snap builds. My packages update fully automatically every two hours, so deb releases are never more than two hours behind the official release. It's entirely hands-off.

Sorry for not catching up.
Is your package referring to the one distributed on zotero.org? Or any link about where this package is provided.

**Comment by retorquere** [comment]
My debs are distributed from https://zotero.retorque.re/file/apt-package-archive

But it is my understanding that the snap store does the distribution? In which case my build would just assure the store is up to date.

**Comment by retorquere** [comment]
Setup for this repo is described on https://github.com/retorquere/zotero-deb

**Comment by retorquere** [comment]
@extraymond can you walk me through the process of getting a snap built/published?

**Comment by retorquere** [comment]
Reading through the discussion above I'm not sure my pipeline is a good fit:

* user-facing work: with the debs, after the one-time repo setup, the user can use the platform native tools for installs and upgrades, and there's no further command line work required or extra plugins to install into the install infra, and that does seem to be required in what I read above. I can't do support for snap installs, as I don't use snap, and I don't really want to put out (more) software that I would struggle to support.
* wrt snap store: my deb builds are fully hands off and don't really work well for stuff that requires manual intervention for stuff like reviews, and also here, I wouldn't be able to answer any questions the reviewers might have, as I don't use snaps. Also not keen to enter the arena of gatekeepers between me my users. If there's an urgent fix I don't want to be held to somebody else's schedules or have to argue priorities with them. Very negative experiences in the past with the Mozilla review process.

**Comment by retorquere** [comment]
> Automatic delta updates, I will always get the latest version from developer.

While the debs don't do deltas, the debs also get you the latest version within two hours of the formal release. 

> Safe revert system, if the new release have bugs, users can snap revert and get the last working version. No need to remove and install older version.

Downgrading isn't as simple as that, admittedly, but not hard either with the debs. Also, this should only be used for emergencies. Otherwise it is in everyone's benefit that problems with the current version get reported. 

> App data are backuped automatically too.

The debs don't offer this, and this would certainly be a benefit. 

> Channel selection, some times I like to try features in the beta version, so I can help discover bugs or report user experience to help devs make better software.

The debs offer this. You can install the beta and the release side by side without issue (other than library incompatibility between them).

**Comment by spipau** [comment]
> In case you weren't aware, there is an up-to-date Zotero flatpak on flathub. https://flathub.org/apps/details/org.zotero.Zotero This may be sufficient and if so this issue could be closed.

I am using it and works great! [It is currently "Unverified"](https://flathub.org/apps/org.zotero.Zotero) and it would be great if Zotero could take over ownership and verify the releases. Also, the official website still offers a download instead of forwarding the users to Flathub.
